app/src/main/java/org/xbmc/kore/ui/sections/localfile/LocalPictureListFragment.java

```diff
@@ -0,0 +1,766 @@
+    /**
+     * Handler on which to post RPC callbacks
+     */
+    private Handler callbackHandler = new Handler();
+
+    String mediaType = Files.Media.FILES;
+    ListType.Sort sortMethod = null;
+    String parentDirectory = null;
+    int playlistId = PlaylistType.MUSIC_PLAYLISTID;             // this is the ID of the music player
+//    private MediaPictureListAdapter adapter = null;
+    boolean browseRootAlready = false;
+    LocalPictureListFragment.FileLocation loadOnVisible = null;
+
+    ArrayList<FileLocation> rootFileLocation = new ArrayList<>();
+    Queue<FileLocation> mediaQueueFileLocation = new LinkedList<>();
+
+    @Override
+    public void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
+        outState.putString(MEDIA_TYPE, mediaType);
+        outState.putParcelable(SORT_METHOD, sortMethod);
+        try {
+            outState.putParcelableArrayList(PATH_CONTENTS, (ArrayList<FileLocation>) ((MediaPictureListAdapter) getAdapter()).getFileItemList());
+        } catch (NullPointerException npe) {
+            // adapter is null probably nothing was save in bundle because the directory is empty
+            // ignore this so that the empty message would display later on
+        }
+        outState.putParcelableArrayList(ROOT_PATH_CONTENTS, rootFileLocation);
+        outState.putBoolean(ROOT_VISITED, browseRootAlready);
+    }
+
+    @Override
+    protected RecyclerViewEmptyViewSupport.OnItemClickListener createOnItemClickListener() {
+        return new RecyclerViewEmptyViewSupport.OnItemClickListener() {
+            @Override
+            public void onItemClick(View view, int position) {
+                handleFileSelect(((MediaPictureListAdapter) getAdapter()).getItem(position));
+            }
+        };
+    }
+
+    @Override
+    protected RecyclerView.Adapter createAdapter() {
+        return new MediaPictureListAdapter(getActivity(), R.layout.grid_item_picture);
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        View root = super.onCreateView(inflater, container, savedInstanceState);
+        Bundle args = getArguments();
+        FileLocation rootPath = null;
+
+
+        if (ContextCompat.checkSelfPermission(this.getActivity(), Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                != PackageManager.PERMISSION_GRANTED) {
+            // Permission to read files was not granted,
+            // prompt the user for storage access permissions:
+            ActivityCompat.requestPermissions(getActivity(),
+                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
+                    42);
+        }
+
+        try {
+            if (http_app == null) {
+                HttpApp.setActivity(getActivity());
+                HttpApp.setPort(8080);
+                http_app = HttpApp.getInstance();
+            } else {
+                http_app.start();
```

Comment:
If there's a way to start it, should there be a way to stop it (and avoid consuming resources)?
And i don't get this, in the previous branch shouldn't it also be started?

Reply:
Stopping is a bit more complicated, because Kodi may still be reading from the HTTP server. For example, if you are playing a local video or MP3.

The HTTP server should only be destroyed when you are sure Kodi isn't reading anything anymore.

Reply:
Got it, hadn't thought of that.